google/cloud/bigtable/examples/stock_backtesting/backtesting_main.cc

```diff
@@ -0,0 +1,239 @@
+  // For one specific row key, we want to query the |base| and |sample| columns
+  // specified in strategy, and the timestamp of [start_date, end_date].
+  // Use filter combo to implement this.
+
+  // Column filters, only select the targeting columns.
+  std::vector<cbt::Filter> column_filters;
+  std::set<std::string> existing_signal;
+  for (auto const& signal : {strategy.base(), strategy.sample()}) {
+    if (existing_signal.find(signal) == existing_signal.end()) {
+      existing_signal.insert(signal);
+      std::vector<std::string> split = absl::StrSplit(signal, kColumnDelimiter);
+      // In the format of column_family::column_qualifier.
+      if (split.size() != 2) {
+        throw std::runtime_error("Invalid strategy definition.");
+      }
+      column_filters.emplace_back(
+          cbt::Filter::ColumnName(split.front(), split.back()));
+    }
+  }
+
+  cbt::Filter column_filter = cbt::Filter::InterleaveFromRange(
+      column_filters.begin(), column_filters.end());
+
+  // Timestamp filter, only select cells in the target window.
+  auto const start_time = absl::FromCivil(start_date, absl::UTCTimeZone());
+  auto const end_time = absl::FromCivil(end_date, absl::UTCTimeZone());
+  auto const start_milli =
+      absl::ToChronoMilliseconds(start_time - absl::UnixEpoch());
+  auto const end_milli =
+      absl::ToChronoMilliseconds(end_time - absl::UnixEpoch());
+  cbt::Filter timestamp_filter =
+      cbt::Filter::TimestampRange(start_milli, end_milli);
+
+  return cbt::Filter::Chain(column_filter, timestamp_filter);
+}
+
+void CalculateProfit(
+    std::unordered_map<std::string, std::map<absl::CivilDay, double>> const&
+        time_series,
+    cbt::examples::Strategy const& strategy) {
+  auto base_it = time_series.find(strategy.base())->second.begin();
+  auto sample_it = time_series.find(strategy.sample())->second.begin();
+  ++sample_it;
+
+  double shares = 0.0, wallet = 0.0;
+  while (sample_it != time_series.find(strategy.sample())->second.end()) {
+    double const sample_price = sample_it->second;
+    double const base_price = base_it->second;
+    double const change = (sample_price - base_price) / base_price;
+    for (auto const& condition : strategy.conditions()) {
+      if (condition.threshold() > 0.0 && change > condition.threshold()) {
+        // Price goes up; buy in.
+        shares += condition.moneyin() / sample_price;
+        wallet -= condition.moneyin();
+      } else if (condition.threshold() < 0.0 &&
+                 change < condition.threshold()) {
+        // Pirce goes down; sell out.
+        shares -= condition.moneyin() / sample_price;
+        wallet += condition.moneyin();
+      }
+    }
+
+    ++sample_it;
+    ++base_it;
+  }
+
+  // At the last day of the backtesting period, we evaluate the total value of
+  // our investment, and calculate the profit.
+  std::cout << "Shares in hand: " << shares << " @ " << base_it->second
+            << std::endl;
```

Comment:
The comment I had about not using `std::endl` (unless you require a flush, which never appears to be the case) still stands.

Reply:
Done.